server/src/org/apache/cloudstack/agent/mslb/AgentMSLBServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+import com.cloud.hypervisor.Hypervisor;
+import com.cloud.resource.ResourceManager;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.naming.ConfigurationException;
+
+import org.apache.cloudstack.agent.mslb.algorithm.AgentMSLBRoundRobinAlgorithm;
+import org.apache.cloudstack.agent.mslb.algorithm.AgentMSLBShuffleAlgorithm;
+import org.apache.cloudstack.agent.mslb.algorithm.AgentMSLBStaticAlgorithm;
+import org.apache.cloudstack.config.ApiServiceConfiguration;
+import org.apache.cloudstack.framework.config.ConfigKey;
+import org.apache.cloudstack.framework.config.Configurable;
+import org.apache.log4j.Logger;
+
+import com.cloud.utils.component.ComponentLifecycleBase;
+import com.cloud.utils.exception.CloudRuntimeException;
+import com.google.common.base.Strings;
+
+public class AgentMSLBServiceImpl extends ComponentLifecycleBase implements AgentMSLB, Configurable {
+    public static final Logger LOG = Logger.getLogger(AgentMSLBServiceImpl.class);
+
+    public static final ConfigKey<String> ConnectedAgentLBAlgorithm = new ConfigKey<>("Advanced", String.class,
+            "connected.agent.mslb.algorithm", "static",
+            "The algorithm to applied on the provided 'host' management server list that is sent to indirect agents. Allowed values are: static, roundrobin and shuffle.",
+            true, ConfigKey.Scope.Global);
+
+    private static Map<String, AgentMSLBAlgorithm> algorithmMap = new HashMap<>();
+
+    @Inject
+    ResourceManager resourceManager;
+    @Inject
+    AgentManager agentManager;
+
+    @Override
+    public List<String> getManagementServerList() {
+        final String msServerAddresses = ApiServiceConfiguration.ManagementServerAddresses.value();
+        if (Strings.isNullOrEmpty(msServerAddresses)) {
+            throw new CloudRuntimeException(String.format("No management server addresses are defined in '%s' setting",
+                    ApiServiceConfiguration.ManagementServerAddresses.key()));
+        }
+
+        return getAgentMSLBAlgorithm().getMSList(Arrays.asList(msServerAddresses.replace(" ", "").split(",")));
+    }
+
+    protected List<HostVO> getOrderedRunningIndirectAgentIds() {
+        List<HostVO> hosts = resourceManager.findDirectlyConnectedHosts();
+        if (hosts != null) {
+            return hosts.stream()
+                    .filter(x -> x.getState().equals(Status.Up) && x.getHypervisorType().equals(Hypervisor.HypervisorType.KVM))
+                    .sorted((x,y) -> Long.compare(x.getId(), y.getId()))
+                    .collect(Collectors.toList());
+        }
+        return null;
+    }
+
+    @Override
+    public void propagateManagementServerListToAgents() {
+        List<HostVO> hosts = getOrderedRunningIndirectAgentIds();
+        if (hosts != null) {
+            for (HostVO h : hosts) {
```

Comment:
@nvazquez I don't think it's a good idea, `resource claim and ownership` is not within the scope of this feature. This method will send command/answers to all the hosts (even those not managed by the `this` management server) and may cause disconnections. In case of multiple management servers, this will cause all the management servers to send multiple requests to all the agents. (note, when you do agent.send(), if the agent is not connected to `this` management server, it will use clustered mgmt server impl to send the request to the forwarding agent which is via other mgmt servers).